File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.cpp
#include<vector>
#include<queue>
using namespace std;

int solution(vector<vector<int> > maps)
{
    int n = maps.size(); // 행
    int m = maps[0].size(); // 열
    vector<vector<int>> visited(n, vector<int>(m, 0)); 
    int dx[4] = {0, 0, 1, -1}; 
    int dy[4] = {1, -1, 0, 0}; 
    queue<pair<int, int>> q; 
    q.push({0, 0}); 
    visited[0][0] = 1; 

    while (!q.empty()) {  
        pair<int, int> now = q.front(); 
        q.pop();
        int x = now.first; 
        int y = now.second; 
        for (int i = 0; i < 4; i++) { 
            int nx = x + dx[i]; 
            int ny = y + dy[i]; 
            if (nx >= 0 && ny >= 0 && nx < n && ny < m) { 
                if (maps[nx][ny] == 1 && visited[nx][ny] == 0) { 
                    visited[nx][ny] = visited[x][y] + 1;
                    q.push({nx, ny}); 
                }
            }
        }
    }
    return visited[n - 1][m - 1] == 0 ? -1 : visited[n - 1][m - 1]; 

}

File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.java
import java.util.*;
class Solution {
    static int[] dx = new int[]{-1, 1, 0, 0};
    static int[] dy = new int[]{0, 0, -1, 1};
    
    public int solution(int[][] maps) {
        int answer = 0;
        int n = maps.length;
        int m = maps[0].length;
        boolean visited[][] = new boolean[n][m];
        Queue<int[]> que = new LinkedList<>();
        que.offer(new int[]{0, 0, 1}); // x, y, distance
        visited[0][0] = true;
        
        while(!que.isEmpty()){
            int[] now = que.poll();
            int x = now[0];
            int y = now[1];
            int dist = now[2];
            
            for(int i = 0 ; i< 4; i++){
                int nx = x + dx[i];
                int ny = y + dy[i];
                
                // 목적지에 도달하면 종료
                if (x == n - 1 && y == m - 1) {
                    return dist;
                }
            
                if (nx >= 0 && ny >= 0 && nx < n && ny < m) {
                    if(maps[nx][ny] == 1 && !visited[nx][ny]){
                        visited[nx][ny] = true;
                        que.offer(new int[] {nx, ny, dist+1});
                    }
                }
            }
        }
        return -1;
    }
}